### PR TITLE
remove read-only attributes on delete for Windows platforms

### PR DESCRIPTION
**Pull request by busterb**
I was looking at https://github.com/rapid7/metasploit-framework/blob/master/lib/msf/core/exploit/file_dropper.rb#L145 and decided to do something about it, teaching Meterpreter how to love / delete files by removing the read-only attribute. We already do it for directories, so files was not much of a stretch.

Doing it portably in non-C meterpreters, I'm not entirely sure of the best approach. What do you think about how I did PHP (other than not having tested it yet) ?

### CODE REVIEW

**Comment by timwr** [comment]
We need to add a meterpreter command for chmod and chown/chgrp too. Currently we have to shell out to do this.

**Comment by busterb** [comment]
ah, thanks!

**Comment by bwatters-r7** [comment]
# Testing on Windows 10x64
## Without Patch:
```

File: php/meterpreter/ext_server_stdapi.php [hunk 1]
```diff
@@ -368,7 +368,11 @@ function stdapi_fs_chdir($req, &$pkt) {
 function stdapi_fs_delete($req, &$pkt) {
     my_print("doing delete");
     $path_tlv = packet_get_tlv($req, TLV_TYPE_FILE_NAME);
-    $ret = @unlink(cononicalize_path($path_tlv['value']));
```
Comment (busterb) [inline php/meterpreter/ext_server_stdapi.php]: cononicalize_path - is this what it is really called? or has delete always been broken in PHP?

Comment (timwr) [reply php/meterpreter/ext_server_stdapi.php]: Oops. Good spot!

Comment (timwr) [reply php/meterpreter/ext_server_stdapi.php]: Oh maybe not. It's a function we define ourselves further up. Typo though

Comment (busterb) [reply php/meterpreter/ext_server_stdapi.php]: fixed anyhow :)

File: python/meterpreter/ext_server_stdapi.py [hunk 2]
```diff
@@ -1220,8 +1220,10 @@ def stdapi_fs_chdir(request, response):
 
 @register_function
 def stdapi_fs_delete(request, response):
-    file_path = packet_get_tlv(request, TLV_TYPE_FILE_NAME)['value']
-    os.unlink(unicode(file_path))
+    file_path = unicode(packet_get_tlv(request, TLV_TYPE_FILE_NAME)['value'])
+    if has_windll:
+        subprocess.call(unicode("attrib.exe -r ") + file_path)
+    os.unlink(file_path)
     return ERROR_SUCCESS, response
```
Comment (bwatters-r7) [inline python/meterpreter/ext_server_stdapi.py]: This does not work in my testing; when I delete a ro file, it calls `stdapi_fs_delete_file`, not `stdapi_fs_delete` (this function).
If I move this syscall of attrib.exe to `stdapi_fs_delete_file` on line 1241, it works great.  Offhand, I'm unclear on the difference between the two, so I'm not sure if it should be moved or copied.

Comment (bwatters-r7) [reply python/meterpreter/ext_server_stdapi.py]: FYI, @bcook-r7 I sent a PR to your branch with the fix, though it is only 2 lines.

**Comment by busterb** [comment]
Merged, thanks @bwatters-r7

**Comment by bwatters-r7** [comment]
# Release Notes
This adds the ability to delete read-only files in c, Python, and PHP meterpreters.

File: php/meterpreter/ext_server_stdapi.php [hunk 3]
```diff
@@ -368,7 +368,11 @@ function stdapi_fs_chdir($req, &$pkt) {
...
+    $path = canonicalize_path($path_tlv['value']);
+    if (is_windows()) {
+        exec("attrib.exe -r ".$path);
+    }
+    $ret = @unlink($path)
```
Comment (busterb) [inline php/meterpreter/ext_server_stdapi.php]: crud, missing a semi-colon. fixed in 96fb621a

**Pull request merged**